### PR TITLE
Added more documentation

### PR DESCRIPTION
**Pull request by RichardLitt**
So far: 
- [x] add

### CODE REVIEW

File: apiary.apib [hunk 1]
```diff
@@ -4,9 +4,41 @@ FORMAT: 1A
...
+    + silent (boolean, optional) - Write no output
+    + progress (boolean, optional) - Stream progress data
+    + trickle (boolean, optional) - Use trickle-dag format for dag generation
+    + only-hash (boolean, optional) - Only chunk and hash - do not write to disk
+    + wrap-with-directory (boolean, optional) - Wrap files with a directory object
+    + hidden (boolean, optional) - Include files that are hidden
+    + chunker (boolean, optional) - chunking algorithm to use
+    + pin (boolean, optional) - Pin this object when adding.  Default true
+
++ Response 200 (text/plain)
```
Comment (RichardLitt) [inline apiary.apib]: I was unable to get "http://127.0.0.1:5001/api/v0/add?arg=/Users/richard/test" to work on Postman, so I honestly am not sure if this is the right response of not. It's what I get in the CLI, though.

File: apiary.apib [hunk 2]
```diff
@@ -4,9 +4,41 @@ FORMAT: 1A
...
+    + recursive (boolean, optional) - Add directory paths recursively
+    + quiet (boolean, optional) - Write minimal output
+    + silent (boolean, optional) - Write no output
+    + progress (boolean, optional) - Stream progress data
+    + trickle (boolean, optional) - Use trickle-dag format for dag generation
+    + only-hash (boolean, optional) - Only chunk and hash - do not write to disk
+    + wrap-with-directory (boolean, optional) - Wrap files with a directory object
+    + hidden (boolean, optional) - Include files that are hidden
+    + chunker (boolean, optional) - chunking algorithm to use
+    + pin (boolean, optional) - Pin this object when adding.  Default true
```
Comment (RichardLitt) [inline apiary.apib:29]: Not sure how to add shortcuts we use here, like -r for --recursive. Should I put them in the desc?

Comment (jbenet) [inline apiary.apib:29]: good question. desc works? not sure. others' thoughts?

Comment (dignifiedquire) [inline apiary.apib:29]: Is there actually a need to have these shortcuts working/documented in the http api? Or should they simply be a bonus specific to the cli?

**Comment by RichardLitt** [comment]
Ok, I've finished going through all of the commands for the first time. Now is the harder bit; which is, namely, figuring out why some don't work in the API at all, where the inconsistencies are in the documentation, and how to clean up and make this, as a whole, better. For now, though, I think this is a massive step forward. Can someone give a brief overview of my work here, point out any glaring issues?

**Comment by RichardLitt** [comment]
This, of course, relates and possibly closes #7.

**Comment by jbenet** [comment]
cc @whyrusleeping @diasdavid @lgierth @mappum for review

File: apiary.apib (hunk 2 above)
Comment (RichardLitt) [inline apiary.apib:29]: I don't think we need them.

**Comment by RichardLitt** [comment]
I have now linted this, and it works in apiary. Whether or not it is 100% accurate is up for debate; but, this means we have something to work with, as far as three columns and apiary goes.

**Comment by dignifiedquire** [comment]
Okay, I will start reviewing this and then we can merge this for now and have any bikeshedding discussions later, as the most important thing for now is to get the IS state documented

File: apiary.apib (hunk 1 above)
Comment (dignifiedquire) [inline apiary.apib]: Path to a file will not work from http api, you have to send the file content via multipart if you want to add something via http

File: apiary.apib [hunk 3]
```diff
@@ -79,8 +116,29 @@ Store data as a raw ipfs block.
...
 
 ## put [GET /block/put]
-TODO
+Stores input as an IPFS block
+
+ipfs block put is a plumbing command for storing raw ipfs blocks.
+It reads from stdin, and <key> is a base58 encoded multihash.
+
++ Parameters
+    + arg (string, required) - The path to the data to be stored as an IPFS block
```
Comment (dignifiedquire) [inline apiary.apib:191]: I don't think this works, `file paths` do not work on the http api as far as I know, or at least there is also an alternative way of uploading the data directly

File: apiary.apib [hunk 4]
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
...
+
+**Examples:**
+
+Get the value of the 'datastore.path' key:
+
+    `ipfs config datastore.path`
+
+Set the value of the 'datastore.path' key:
+
+    `ipfs config datastore.path ~/.ipfs/datastore`
```
Comment (dignifiedquire) [inline apiary.apib]: These examples should not be here, as they document the cli

File: apiary.apib [hunk 5]
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
 
 # Group config
 
+## config [GET /config{?arg1,arg2}{&bool,json}]
+
+ipfs config controls configuration variables. It works
+much like 'git config'. The configuration values are stored in a config
```
Comment (dignifiedquire) [inline apiary.apib]: please use ticks instead of '

File: apiary.apib [hunk 6]
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
...
+
+Set the value of the 'datastore.path' key:
+
+    `ipfs config datastore.path ~/.ipfs/datastore`
+
++ Parameters
+    + arg1: "Addresses.API" (string, required) - The key of the config entry
+    + arg2 (string, required) - The value to set the config entry to
+    + bool (boolean, optional) - Set a boolean
+    + json (boolean, optional) - Parse stringified JSON
```
Comment (dignifiedquire) [inline apiary.apib:405]: It's not documented how you would actually send json

File: apiary.apib [hunk 7]
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
...
-+ Response 200 (text/plain)
+## replace [GET /config/replace{?arg}]
+Replaces the config with <file>
+
+Make sure to back up the config file first if neccessary, this operation
+can't be undone.
+
++ Parameters
+
+    + arg (string, required) - The file to use as the new config
```
Comment (dignifiedquire) [inline apiary.apib:437]: Will this work with a file path, or do you need to attach the actual content in the request?

File: apiary.apib [hunk 8]
```diff
@@ -244,41 +305,301 @@ Lists all available commands (and subcommands) that are available.
...
+over the network. Most applications that use IPFS will do so by
+communicating with a daemon over the HTTP API. While the daemon is
+running, calls to `ipfs` commands will be sent over the network to
+the daemon.
+
+The daemon will start listening on ports on the network, which are
+documented in (and can be modified through) `ipfs config Addresses`.
+For example, to change the `Gateway` port:
+
+    ipfs config Addresses.Gateway /ip4/127.0.0.1/tcp/8082
```
Comment (dignifiedquire) [inline apiary.apib]: Again please no cli examples in here

File: apiary.apib [hunk 9]
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
...
+IPNS is a PKI namespace, where names are the hashes of public keys, and
+the private key enables publishing new (signed) values. In publish, the
+default value of <name> is your own identity public key.
 
-+ Response 200 (text/plain)
+## publish [GET /name/publish{?arg}{&resolve,lifetime,ttl}]
+Publish an object to IPNS
+
++ Parameters
+    + arg (string, required) - IPFS path of the obejct to be published
```
Comment (dignifiedquire) [inline apiary.apib]: typo `object`

File: apiary.apib [hunk 10]
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
...
 + Response 200 (text/plain)
 
-## stat [GET /object/stat]
+## put [GET /object/put{?arg}{&inputenc}]
+Stores input as a DAG object, outputs its key
+
+`ipfs object put` is a plumbing command for storing DAG nodes.
+It reads from stdin, and the output is a base58 encoded multihash.
+
+Data should be in the format specified by the --inputenc flag.
```
Comment (dignifiedquire) [inline apiary.apib]: Please no cli speak like `--flag`

File: apiary.apib [hunk 11]
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
...
+            * unixfs-dir
+
 
 + Response 200 (text/plain)
 
-## put [GET /object/put]
+## patch [GET /object/patch{?arg1,arg2}{&name,ref}]
+Create a new merkledag object based on an existing one
+
+`ipfs object patch <root> <cmd> <args>` is a plumbing command used to
```
Comment (dignifiedquire) [inline apiary.apib]: again, this is cli notation

File: apiary.apib [hunk 12]
```diff
@@ -449,91 +868,370 @@ ipfs.io. TXT "dnslink=/ipfs/QmRzTuh2Lpuz7Gr39stNr6mTFdqAghsZec1JoUnfySUzcy ..."
...
+    + Attributes
+        + `Keys` (object) - Contains multiple hashes
+            + hash (object)
+                - `Type`: "recursive" (string)
+                - `Coun`: 1 (number)
+
+## rm [GET /pin/rm{?arg}{&recursive}]
+Removes the pinned object from local storage.
+
+By default, recursively. Use `-r=false` for direct pins.
```
Comment (dignifiedquire) [inline apiary.apib]: cli notation

File: apiary.apib [hunk 13]
```diff
@@ -563,10 +1261,54 @@ trip latency information.
...
 
+## refs [GET /refs{?arg}{&format,edges,unique,recursive}]
+Lists links (references) from an object
+
+Retrieves the object named by <ipfs-path> and displays the link
+hashes it contains, with the following format:
+
+  <link base58 hash>
+
+Note: list all refs recursively with -r.
```
Comment (dignifiedquire) [inline apiary.apib]: cli notation

File: apiary.apib [hunk 14]
```diff
@@ -563,10 +1261,54 @@ trip latency information.
 
 # Group refs
 
+## refs [GET /refs{?arg}{&format,edges,unique,recursive}]
+Lists links (references) from an object
+
+Retrieves the object named by <ipfs-path> and displays the link
+hashes it contains, with the following format:
+
+  <link base58 hash>
```
Comment (dignifiedquire) [inline apiary.apib]: strange indentation

File: apiary.apib [hunk 15]
```diff
@@ -4,9 +4,42 @@ FORMAT: 1A
...
+Add an object to ipfs.
 
-TODO
+Adds contents of <path> to ipfs. Use -r to add directories.
+Note that directories are added recursively, to form the ipfs
+MerkleDAG. A smarter partial add with a staging area (like git)
+remains to be implemented.
+
++ Parameters
+    + arg (string, required) - The path to a file to be added to IPFS
```
Comment (mappum) [inline apiary.apib:19]: We might want to indicate that there can be multiple `arg` values to add multiple files in one request

File: apiary.apib (hunk 1 above)
Comment (mappum) [inline apiary.apib]: If you need to test adds in Postman, you can use the `form-data` body Content-Type, set a non-empty value for the key, and select 'File' instead of 'Text'.

In curl, you can do:

```
curl -F "file=@/some/file/path" http://localhost:5001/api/v0/add
```

File: apiary.apib [hunk 16]
```diff
@@ -79,8 +116,29 @@ Store data as a raw ipfs block.
         ```
 
 ## put [GET /block/put]
-TODO
+Stores input as an IPFS block
+
+ipfs block put is a plumbing command for storing raw ipfs blocks.
+It reads from stdin, and <key> is a base58 encoded multihash.
```
Comment (mappum) [inline apiary.apib]: Stdin isn't relevant in a HTTP API

File: apiary.apib (hunk 1 above)
Comment (mappum) [inline apiary.apib]: BTW, could be cool to show `curl` commands for all these methods, does apiary have any good places where we could put that?

File: apiary.apib (hunk 15 above)
Comment (ABISprotocol) [inline apiary.apib:19]: @mappum Excuse me, I realize this is OT for here, but just a brief inquiry, do you have a status or roadmap for Mercury?

Comment (mappum) [inline apiary.apib:19]: @ABISprotocol Not sure why you're bringing that up here. My email address is on my profile, feel free to contact me there. :smile:

File: apiary.apib (hunk 1 above)
Comment (RichardLitt) [inline apiary.apib]: There's no option to select File instead of Text if I am using a GET command. If I use a PUT or a POST, there is, but I just get a `403 - Forbidden` back.  Ideas?

Comment (RichardLitt) [inline apiary.apib]: Looked around: I can't see any way to actually have the curl commands in these.

Comment (RichardLitt) [inline apiary.apib]: We could include the curl command directly in the description, like this. I like it! What do you think? cc @Dignifiedquire 

```
# Group add [GET /add{?arg}{&recursive,quiet,silent,progress,trickle,only%2Dhash,wrap%2Dwith%2Ddirectory,hidden,chunker,pin}]
Add an object to ipfs.

Adds contents of <path> to ipfs. Use -r to add directories.
Note that directories are added recursively, to form the ipfs
MerkleDAG. A smarter partial add with a staging area (like git)
remains to be implemented.

#### curl

    curl -F "file=@test/" -r http://localhost:5001/api/v0/add

+ Parameters
    + arg (string, required) - The path to a file to be added to IPFS
...
```

Comment (RichardLitt) [inline apiary.apib]: Looks like this: 
screenshot 2015-12-25 20 50 09

File: apiary.apib (hunk 3 above)
Comment (RichardLitt) [inline apiary.apib:191]: It works with curl. Added example response and curl request.

File: apiary.apib (hunk 7 above)
Comment (RichardLitt) [inline apiary.apib:437]: It works with a file path. Added a `curl` example, and example headers. This should be a 204, not a 200. See #14.

File: apiary.apib (hunk 11 above)
Comment (RichardLitt) [inline apiary.apib]: Yeah, I don't know about this one. I wasn't able to get it to work on Postman. Ought we to remove it?

**Comment by RichardLitt** [comment]
@Dignifiedquire Want to take another pass?

**Comment by RichardLitt** [comment]
Removed 'needs review' pending current changes ongoing after the API talk.

File: apiary.apib (hunk 1 above)
Comment (daviddias) [inline apiary.apib]: > from @RichardLitt: There's no option to select File instead of Text if I am using a GET command. If I use a PUT or a POST, there is, but I just get a 403 - Forbidden back. Ideas?

That would be because no one (i.e Apiary) is expecting for a file to be sent in a GET request and all the other methods are blocked by CORS, if the headers are not set.

File: apiary.apib (hunk 2 above)
Comment (daviddias) [inline apiary.apib:29]: let's make the alias be only available through the CLI :)

Comment (RichardLitt) [inline apiary.apib:29]: :+1:

**Comment by RichardLitt** [comment]
Please do not close or delete this branch. I am using it to stub out future PRs. This PR will not and should not be merged.
